File: Loop/src/EventLoop.cpp
#include "EventLoop.h"
#include <iostream>
#include <stdexcept>

EventLoop::EventLoop() {

}

void EventLoop::initialize() {
    m_chain = new TChain(treeName);
    for (auto inputFile : inputFiles){
        m_chain -> Add(inputFile);
        std::cout << "Added file: " << inputFile << std::endl;
    }

    m_data = new Data(m_chain);
    m_data->initialize();

    for(auto algorithm : algorithms){
        algorithm->initialize(m_data);
    }



}

void EventLoop::execute(){

 if(!m_chain) {
   throw std::runtime_error("Calling execute while the wvent loop was not initialized.");
 }

 for (Long64_t i=0; i<m_chain->GetEntries(); i++){
     m_chain->GetEntry(i);
     for(auto algorithm : algorithms){
         algorithm->execute();
     }

 }

}

void EventLoop::finalize() {
    delete m_chain;
    m_data->finalize();
    for (auto algorithm : algorithms){
        algorithm->finalize();
    }
}

File: Loop/EventLoop.py
#!/usr/bin/env python
from ROOT import gSystem

# 加载共享库，注意库名称可能需要完整路径或库名前缀（在 macOS 下通常为 libloop.dylib）
gSystem.Load("build/libloop.dylib")

# 从共享库中导入 EventLoop 类
from ROOT import EventLoop, Algorithm, TH1D

# 创建 EventLoop 类的实例
eventLoop = EventLoop()

eventLoop.treeName = "NOMINAL"

eventLoop.inputFiles.push_back("../Rawdata/ggH.root")


alg = Algorithm()
alg.h_ditau_m = TH1D("ditau_m", ";#it{m}_{#tau#tau} [GeV];Events", 30, 0, 300)
alg.h_ditau_m.Sumw2()

eventLoop.algorithms.push_back( alg )



# 初始化事件循环（例如：构建 TChain 并加载所有文件）
eventLoop.initialize()

# 运行事件循环（遍历事件并处理）
eventLoop.execute()

# Create a new ROOT file
from ROOT import TFile
f = TFile.Open("histograms.root", "recreate")
f.cd()

alg.h_ditau_m.Write()
f.Close()

# 结束事件循环（例如：保存输出文件）
eventLoop.finalize()
